Add custom field coders for auto serialization

alfonsogarciacaro opened the pull request:
**UPDATE**: The PR has been repurposed, please see the [comment below](https://github.com/thoth-org/Thoth.Json/pull/63#issuecomment-616717880).

This is a proposal to add default values for fields in auto decoded records.

In my current project we're using a NoSql database and we're using auto coders when reading/writing and exchanging data with the frontend. However it often happens that we need to add an extra field to some documents and migrating the db is a bit cumbersome, the only thing we need is using a default value whenever we read the document from the db and write it back. When adding the field to the record type, Thoth fails deserialization if it doesn't find the field in the json. There are two solutions for this:

- Make the field optional, but this forces us to always handle the None case, even if the field is supposed to be required from now on.
- Use a manual decoder. This is what we've been doing but it's adding a lot of boilerplate to the code because we have some documents with many fields, and whenever we want to add a new one, we need to write a full manual decoder (and encoder, even if we don't need it) and we could make mistakes when doing it.

With this PR, we only need to do something like this:

```fsharp
open Fable.Core.Experimental

let extraCoders =
    Extra.empty
    |> Extra.withInt64    
    |> Extra.withDefaultField<FileRequestDTO>
        (nameofLambda(fun x -> x.OrganizationalType)) FileRequestDTO.DefaultOrganization
```

I've made the modification for my project and it's working fine, but I had to duplicate a lot of code from Thoth.Json because most of the methods used for Auto decoding are private. So it'd be great to merge the change into the library. The PR is incomplete as it doesn't include tests, nor the .NET version. If the approach is acceptable and there's a desire to merge it, I will complete it.

MangelMaxime commented:
Hello @alfonsogarciacaro,

can't you use a manual decoder to decode the property that causes the problem and pass an auto-generated decoder to use for the data?

Something like that:

```fs
type Container<'T> =
    {
        Data : 'T
        Id : int
    }

    static member Decoder (dataDecoder : Decoder<'T>)=
        Decode.object (fun get ->
            {
                Data = get.Required.Raw dataDecoder
                Id = get.Required.Field "id" Decode.int
            }
        )
```

If this doesn't work, I think we can add the feature you want to Thoth.Json.

alfonsogarciacaro commented:
Not sure if I understand your proposal but yes, as commented above I was using manual decoders but things became complicated very quickly. To illustrate this I'm showing the code we were using before. I'm including only the biggest type but there were more:

```fsharp
let mutable extraCoders = Extra.empty

// These are helpers we use in the project to generate the coders automatically
let inline decoder<'T> = Decode.Auto.generateDecoderCached<'T>(caseStrategy=CamelCase, extra=extraCoders)
let inline encoder<'T> = Encode.Auto.generateEncoderCached<'T>(caseStrategy=CamelCase, extra=extraCoders)

// This extension was used to easily give a default value to some properties in manual decoders
type Decode.IOptionalGetter with
    member getter.FieldOr fieldName decoder defaultValue =
        match getter.Field fieldName decoder with
        | Some v -> v
        | None -> defaultValue

type ComplianceGuideDTO with
    static member Encode(x: ComplianceGuideDTO) =
        Encode.object [
            "id", Encode.string x.Id
            "companyId", Encode.string x.CompanyId
            "userId", Encode.string x.UserId
            "auditTrail", encoder x.AuditTrail
            "dpoRegistration", encoder x.DpoRegistration
            "dpoRegistrationFiles", encoder x.DpoRegistrationFiles
            "privacyPolicy", encoder x.PrivacyPolicy
            "privacyPolicyFiles", encoder x.PrivacyPolicyFiles
            "ancillaryDataProcessing", encoder x.AncillaryDataProcessing
            "ancillaryDataProcessingFiles", encoder x.AncillaryDataProcessingFiles
            "processCatelogue", encoder x.ProcessCatelogue
            "iTSecurityConcept", encoder x.ITSecurityConcept
            "iTSecurityConceptFiles", encoder x.ITSecurityConceptFiles
            "websitePrivacyPolicy", encoder x.WebsitePrivacyPolicy
            "websitePrivacyPolicyFiles", encoder x.WebsitePrivacyPolicyFiles
            "thirdCountryTransfer", encoder x.ThirdCountryTransfer
            "thirdCountryTransferFiles", encoder x.ThirdCountryTransferFiles
            "deletionPeriod", encoder x.DeletionPeriod
            "deletionPeriodFiles", encoder x.DeletionPeriodFiles
            "toms", encoder x.Toms
            "tomsFiles", encoder x.TomsFiles
            "individualRights", encoder x.IndividualRights
            "dataBreach", encoder x.DataBreach
        ]    
    static member Decoder: Decoder<ComplianceGuideDTO> =         
        Decode.object (fun get ->
            { Id = get.Required.Field "id" Decode.string
              CompanyId = get.Required.Field "companyId" Decode.string
              UserId = get.Required.Field "userId" Decode.string
              AuditTrail = get.Required.Field "auditTrail" decoder
              DpoRegistration = get.Required.Field "dpoRegistration" decoder
              DpoRegistrationFiles = get.Required.Field "dpoRegistrationFiles" decoder
              PrivacyPolicy = get.Required.Field "privacyPolicy" decoder
              PrivacyPolicyFiles = get.Required.Field "privacyPolicyFiles" decoder
              AncillaryDataProcessing = get.Required.Field "ancillaryDataProcessing" decoder
              AncillaryDataProcessingFiles = get.Required.Field "ancillaryDataProcessingFiles" decoder
              ProcessCatelogue = get.Required.Field "processCatelogue" decoder
              ITSecurityConcept = get.Required.Field "iTSecurityConcept" decoder
              ITSecurityConceptFiles = get.Required.Field "iTSecurityConceptFiles" decoder
              WebsitePrivacyPolicy = get.Required.Field "websitePrivacyPolicy" decoder
              WebsitePrivacyPolicyFiles = get.Required.Field "websitePrivacyPolicyFiles" decoder
              ThirdCountryTransfer = get.Required.Field "thirdCountryTransfer" decoder
              ThirdCountryTransferFiles = get.Required.Field "thirdCountryTransferFiles" decoder
              DeletionPeriod = get.Required.Field "deletionPeriod" decoder
              DeletionPeriodFiles = get.Required.Field "deletionPeriodFiles" decoder
              Toms = get.Optional.FieldOr "toms" decoder { OrganizationalProcessDTO.State = 0}
              TomsFiles = get.Optional.FieldOr "tomsFiles" decoder []
              IndividualRights = get.Required.Field "individualRights" decoder
              DataBreach = get.Required.Field "dataBreach" decoder
            }
        )

extraCoders <-
    Extra.empty
    |> Extra.withInt64
    // Custom decoder to give a default value to fields
    // not present in the objects stored in the db
    |> Extra.withCustom ComplianceGuideDTO.Encode ComplianceGuideDTO.Decoder
```

This works but has several issues:

- We need to write the full manual decoder for the whole type even if we only need default values for one or two properties.
- Because the `Extra.withCustom` method needs also the encoder we need to write it too.
- Because in the manual coders we need to call automatic encoders there's a cyclic dependency that we need to solve by making the `extraCoders` mutable (and they must be also public because of the inlined helpers) and initialize them at the bottom of the file, which is not very nice (using `module rec` gives warnings here).
- When writing the name of the properties in the manual coders we need to be sure we don't make typos and we're using the same case strategy as in the `decoder` and `encoder` helpers.

As mentioned above, with this proposal the code becomes much simpler:

```fsharp
let extraCoders =
    Plus.Extra.empty
    |> Plus.Extra.withInt64    
    |> Plus.Extra.withDefaultField<ComplianceGuideDTO>
        (nameofLambda(fun x -> x.Toms)) { OrganizationalProcessDTO.State = 0}
    |> Plus.Extra.withDefaultField<ComplianceGuideDTO>
        (nameofLambda(fun x -> x.TomsFiles)) []

let inline decoder<'T> = Plus.Decode.Auto.generateDecoderCached<'T>(caseStrategy=CamelCase, extra=extraCoders)
let inline encoder<'T> = Plus.Encode.Auto.generateEncoderCached<'T>(caseStrategy=CamelCase, extra=extraCoders)
```

The main issue is probably to decide if this a feature for a very specific case and shouldn't go into the library, or whether it can benefit other users.

MangelMaxime commented:
Well, my proposition was if your type structure was having a wrapper type and not everything flattens. But it doesn't seem to be the case.

This is indeed a specific case but I do have something a similar feature on my backend at work.

Indeed it should go into a library but it would require you to duplicate the reflection code heavily so it is not a good thing.

We can add it as a new feature of Thoth.Json :)

alfonsogarciacaro commented:
I've bee trying to foresee future requests for the auto serialization so we don't have to make the code too complicated with too many options. And at the end I've thought we could add custom field coders when you need some more granular tuning for some specific cases (using a default value as in the original intention of this PR, hiding sensitive data like user info in the json, etc). Please let me know what you think. I've added a couple of tests in ExtraCoders that could be used as an example.

As a bonus, I've also added support for mutable dictionaries and hashsets in auto serialization, which was also needed in one of my projects (I solved it with custom coders, but it's also tedious because you need to add them for each type parameter).

MangelMaxime commented:
Thank you @alfonsogarciacaro, I think the change is good and the new functionality can be handful.

The only thing I worry about is that people code will become a bunch more complex when using the new functionalities but I can't think of another way to solve the problem. Except for the manual decoders but I think I am in the minority who like them ^^

Writing the documentation for all that will be fun :D

I am now looking at Thoth.Json.Net version.

MangelMaxime commented on src/Encode.fs:
```diff
@@ -472,38 +504,25 @@ If you can't use one of these types, please pass an extra encoder.
+                    || fullname = typedefof<Set<string>>.FullName
+                    || fullname = typedefof<HashSet<string>>.FullName then
                     // Disable seq support for now because I don't know how to implements to on Thoth.Json.Net
                     // || fullname = typedefof<obj seq>.FullName then
```
@alfonsogarciacaro Can I abuse a bit of your presence and ask you if you know-how to implement seq support in Thoth.Json.Net ? 🙏 😄 

I was able to do it for Thoth.Json but unfortunately not for Thoth.Json.Net...

alfonsogarciacaro replied:
I saw the comment and did try to implement support for `seq` too but as you said it wasn't straightforward to do in .Net (don't remember exactly what was the issue) and I quickly gave up to avoid spending too much time with the PR. But I'll give it another go :)

alfonsogarciacaro commented:
Yes, I agree the API is not super nice but I tried to make it flexible enough so we don't have to touch the code much more. Most users shouldn't need to know about it, and it's possible to create other helpers on top of it. For example, to make the default values as originally intended:

```fsharp
let inline withDefaultValue<'T> (fieldName: string) (defaultValue: obj) (extra: ExtraCoders) =
    extra |> __withCustomFieldDecoderAndKey typeof<'T>.FullName fieldName (fun path value ->
        match value with
        | Some _ -> UseAutoDecoder
        | None -> UseOk defaultValue)
```

I forgot to extend the documentation but maybe it can remain as a "hidden feature" for now until we test it in a couple of projects in case we see something needs to be changed.

MangelMaxime commented:
Yes, I will go with a beta release indeed before releasing in the wild. 

I just need to prepare all the project who use Thoth.Json(.Net) as a dependency to be released with a new beta too.
